src/core/regimes.rkt

```diff
@@ -175,127 +175,119 @@
+       ;; the current alt-error-sum is less then previous
+       (when (or (not best-alt-idx) (fl< current-error best-alt-cost))
+        ;; update best cost and best index
+        (flvector-set! best-alt-costs prev-split-idx current-error)
+        (vector-set! best-alt-idxs prev-split-idx alt-idx)))))
+   ;; We have now have the index of the best alt and it's error up to our 
+   ;; current point-idx.
+   ;; Now we compare against our current best saved in the 3 vectors above
+   (for ([prev-split-idx (in-range 0 point-idx)]
+         [r-error-sum (in-flvector result-error-sums)]
+         [best-alt-idx (in-vector best-alt-idxs)]
+         [best-alt-cost (in-flvector best-alt-costs)]
+         [can-split (in-vector can-split-vec 1)]
+         #:when can-split)
+     ;; Re compute the error sum for a potential better alt
+     (define alt-error-sum (fl+ r-error-sum best-alt-cost min-weight))
+     ;; Check if the new alt-error-sum is better then the current
+     (define set-cond
+      ;; give benefit to previous best alt
+      (cond [(fl< alt-error-sum current-alt-error) #t]
+            ;; Tie breaker if error are the same favor first alt
+            [(and (fl= alt-error-sum current-alt-error)
+                  (> current-alt-idx best-alt-idx)) #t]
+            ;; Tie breaker for if error and alt is the same
+            [(and (fl= alt-error-sum current-alt-error)
+                  (= current-alt-idx best-alt-idx)
+                  (> current-prev-idx prev-split-idx)) #t]
+            [else #f]))
+      (when set-cond
+       (set! current-alt-error alt-error-sum)
+       (set! current-alt-idx best-alt-idx)
+       (set! current-prev-idx prev-split-idx)))
+   (flvector-set! result-error-sums point-idx current-alt-error)
+   (vector-set! result-alt-idxs point-idx current-alt-idx)
+   (vector-set! result-prev-idxs point-idx current-prev-idx))
 
-  ;; start at (- num-points 1)
-  ;; if num-points we are done
-  ;; traversing and then reversing is bad
-  (define (build-list current-cand)
-    (cond 
-      [(not(= (cand-prev-idx current-cand) num-points))
-        (cons (si (cand-idx current-cand) (cand-point-idx current-cand))
-               (build-list (vector-ref fixed-final (cand-prev-idx current-cand))))]
-      [else 
-        (cons (si (cand-idx current-cand) (cand-point-idx current-cand)) (list))]))
-  (reverse (build-list (vector-ref fixed-final (- num-points 1)))))
+  ;; Loop over results vectors in reverse and build the output split index list
+  (define next number-of-points)
+  (define split-idexs #f)
+  (for ([i (in-range (- number-of-points 1) -1 -1)]
+        #:when (= (+ i 1) next))
+   (define alt-idx (vector-ref result-alt-idxs i))
+   (define split-idx (vector-ref result-prev-idxs i))
+   (set! next (+ split-idx 1))
+   (set! split-idexs (cond
+    [(false? split-idexs) (cons (si alt-idx number-of-points) '())]
+    [else (cons (si alt-idx (+ i 1)) split-idexs)])))
+ split-idexs)
```

Comment:
Really wanted make this a `(for/list`

Reply:
This is pretty clever, using a for loop instead of a recursive function.

Reply:
Thanks, wasn't my intension was just trying to clean it up. Though Is every cons an allocation here? I wonder if I can figure out the number of split points that the output needs and use a vector instead of a list. Though this is outside the core loop and probably best to work on other things.

Reply:
`cons` allocates (it must allocate space to hold the car & cdr, I believe there's also a 64-bit tag so I think each `cons` allocates 196 bytes.

I'm sure we could do the optimization you suggest but TBH this is typically doing, like, four allocations, it's probably not perf critical. At this point if we're going to do more work on regimes we really do have to profile more carefully to figure out where we spend the time.